# Filtering data

In the last section we looked at how to act on entire columns at once. For example when we did:

```python
tips["total_bill"] * 100
```

it applied the multiplication to every row, multiplying each number by 100.

Sometimes we don't want to have to deal with entire columns at once, we might only want to grab a subset of the data and look in just that part. For example, with the tips data, we might think that the day of the week will affect the data so we just want to grab the data for Saturdays.

In Pandas there are two steps to asking a question like this.

1. create a *filter* which describes the question you want to ask
2. *apply* that filter to the data to get just the bits you are interested in

You create a filter by performing some operation on your `DataFrame` or a column within it. To ask about only those rows which refer to Saturday, you grab the `day` column and compare it to `"Sat"`:

In [1]:
import pandas as pd
tips = pd.read_csv("./data/tips.csv")

In [2]:
tips

total_bill   tip   day    time  size
0         16.99  0.71   Sun  Dinner     2
1         10.34  1.16   Sun  Dinner     3
2         21.01  2.45   Sun  Dinner     3
3         23.68  2.32   Sun  Dinner     2
4         24.59  2.53   Sun  Dinner     4
..          ...   ...   ...     ...   ...
239       29.03  4.14   Sat  Dinner     3
240       27.18  1.40   Sat  Dinner     2
241       22.67  1.40   Sat  Dinner     2
242       17.82  1.22   Sat  Dinner     2
243       18.78  2.10  Thur  Dinner     2

[244 rows x 5 columns]

In [3]:
tips["day"] == "Sat"

0      False
1      False
2      False
3      False
4      False
       ...  
239     True
240     True
241     True
242     True
243    False
Name: day, Length: 244, dtype: bool

This has created a filter object (sometimes called a *mask* or a *boolean array*) which has `True` set for the rows where the day is Saturday and `False` elsewhere.

We could save this filter as a variable:

In [4]:
sat_filter = (tips["day"] == "Sat")

We can use this to filter the `DataFrame` as a whole. `tips["day"] == "Sat"` has returned a `Series` containing booleans. Passing it back into `tips` as an indexing operation will use it to filter based on the `day` column, only keeping those rows which contained `True` in the filter:

In [5]:
tips[sat_filter]

total_bill   tip  day    time  size
19        20.65  2.34  Sat  Dinner     3
20        17.92  2.86  Sat  Dinner     2
21        20.29  1.92  Sat  Dinner     2
22        15.77  1.56  Sat  Dinner     2
23        39.42  5.31  Sat  Dinner     4
..          ...   ...  ...     ...   ...
238       35.83  3.27  Sat  Dinner     3
239       29.03  4.14  Sat  Dinner     3
240       27.18  1.40  Sat  Dinner     2
241       22.67  1.40  Sat  Dinner     2
242       17.82  1.22  Sat  Dinner     2

[87 rows x 5 columns]

Notice that it now says that the table only has 87 rows, down from 244. However, the index has been maintained. This is because the row labels are connected to the row, they're not just row numbers.

It is more common to do this in one step, rather than creating and naming a filter object. So the code becomes:

In [6]:
tips[tips["day"] == "Sat"]

total_bill   tip  day    time  size
19        20.65  2.34  Sat  Dinner     3
20        17.92  2.86  Sat  Dinner     2
21        20.29  1.92  Sat  Dinner     2
22        15.77  1.56  Sat  Dinner     2
23        39.42  5.31  Sat  Dinner     4
..          ...   ...  ...     ...   ...
238       35.83  3.27  Sat  Dinner     3
239       29.03  4.14  Sat  Dinner     3
240       27.18  1.40  Sat  Dinner     2
241       22.67  1.40  Sat  Dinner     2
242       17.82  1.22  Sat  Dinner     2

[87 rows x 5 columns]

This has given us back our subset of data as another `DataFrame` which can used in exactly the same way as the previous one (further filtering, summarising etc.).

### Exercise 1

- Select the data for only Thursdays.
- Calculate the mean of the `tip` column for Thursdays
- Compare this with the mean of the `tip` column for Saturdays


In [7]:
thurs = tips[tips["day"] == "Thur"]
thurs

total_bill   tip   day    time  size
77        27.20  2.80  Thur   Lunch     4
78        22.76  2.10  Thur   Lunch     2
79        17.29  1.90  Thur   Lunch     2
80        19.44  2.10  Thur   Lunch     2
81        16.66  2.38  Thur   Lunch     2
..          ...   ...   ...     ...   ...
202       13.00  1.40  Thur   Lunch     2
203       16.40  1.75  Thur   Lunch     2
204       20.53  2.80  Thur   Lunch     4
205       16.47  2.26  Thur   Lunch     3
243       18.78  2.10  Thur  Dinner     2

[62 rows x 5 columns]

In [8]:
thurs["tip"].mean()

1.9398387096774192

In [9]:
tips[tips["day"] == "Sat"]["tip"].mean()

2.095402298850575

## Other filters

As well as filtering with the `==` operator (which only checks for exact matches), you can do other types of comparisons. Any of the standard Python comparisons will work (i.e. `==`, `!=`, `<`, `<=`, `>`, `>=`).

To grab only the rows where the total bill is less than £8 we can use `<`:

In [10]:
tips[tips["total_bill"] < 8]

total_bill   tip   day    time  size
67         3.07  0.70   Sat  Dinner     1
92         5.75  0.70   Fri  Dinner     2
111        7.25  0.70   Sat  Dinner     1
149        7.51  1.40  Thur   Lunch     2
172        7.25  3.60   Sun  Dinner     2
195        7.56  1.01  Thur   Lunch     2
218        7.74  1.01   Sat  Dinner     2

### Exercise 2

Filter the data to only include parties of 5 or more people.



In [11]:
tips[tips["size"] >= 5]

total_bill   tip   day    time  size
125       29.80  2.94  Thur   Lunch     6
141       34.30  4.69  Thur   Lunch     6
142       41.19  3.50  Thur   Lunch     5
143       27.05  3.50  Thur   Lunch     6
155       29.85  3.60   Sun  Dinner     5
156       48.17  3.50   Sun  Dinner     6
185       20.69  3.50   Sun  Dinner     5
187       30.46  1.40   Sun  Dinner     5
216       28.15  2.10   Sat  Dinner     5

## Combining filters

If you want to apply multiple filters, for example to select only "Saturdays with small total bills" you can do it in one of two different ways. Either split the question into multiple steps, or ask it all at once.

Let's do it multiple steps first since we already have tools we need for that:

In [12]:
sat_tips = tips[tips["day"] == "Sat"]  # First grab the Saturday data and save it as a variable
sat_tips[sat_tips["total_bill"] < 8]  # Then act on the new DataFrame as use it as before

total_bill   tip  day    time  size
67         3.07  0.70  Sat  Dinner     1
111        7.25  0.70  Sat  Dinner     1
218        7.74  1.01  Sat  Dinner     2

Or, you can combine the questions together using the `&` operator with a syntax like:

```python
df[(filter_1) & (filter_2)]
```

so in our case filter 1 is `tips["day"] == "Sat"` and filter 2 is `tips["total_bill"] < 8` so it becomes:

In [13]:
tips[(tips["day"] == "Sat") & (tips["total_bill"] < 8)]

total_bill   tip  day    time  size
67         3.07  0.70  Sat  Dinner     1
111        7.25  0.70  Sat  Dinner     1
218        7.74  1.01  Sat  Dinner     2

If you want to do an "or" operation, then instead of `&` you can use `|`.

### Exercise 3

Filter the data to only include parties of 4 or more people which happened at lunch time.

Hint: The `size` and `time` columns are what you want to use here.


In [14]:
tips[(tips["size"] >= 4) & (tips["time"] == "Lunch")]

total_bill   tip   day   time  size
77        27.20  2.80  Thur  Lunch     4
85        34.83  3.62  Thur  Lunch     4
119       24.08  2.04  Thur  Lunch     4
125       29.80  2.94  Thur  Lunch     6
141       34.30  4.69  Thur  Lunch     6
142       41.19  3.50  Thur  Lunch     5
143       27.05  3.50  Thur  Lunch     6
197       43.11  3.50  Thur  Lunch     4
204       20.53  2.80  Thur  Lunch     4

## DataFrame indexing

When we use the square bracket syntax on a `DataFrame` directly there are a few different types of object that can be passed:

<dl>
<dt>A single string</dt>
<dd>This will select a single column form the <code>DataFrame</code>, returning a <code>Series</code> object.</dd>
<dt>A list of strings</dt>
<dd>This will select those columns by name, returning a <code>DataFrame</code>.</dd>
<dt>A filter (a <code>Series</code> of <code>True</code>/<code>False</code>)</dt>
<dd>This will filter the table as a whole, returning a <code>DataFrame</code> with only the rows matching <code>True</code>included.</dd>
</dl>

These are provided as shortcuts as they are the most common operations to do on a `DataFrame`. This is why some of them operate on columns and other on rows.

If you want to be expicit about which axis you are acting on, you can pass these same types of objects to the `.loc[rows, columns]` attribute with one argument per axis. This means that

```python
tips[sat_filter]
```

is equivalent to
```python
tips.loc[sat_filter]
```

and that
```python
tips["size"]
```

is equivalent to
```python
tips.loc[:, "size"]
```

The full set of rules for [`DataFrame.loc` are in the documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html).

### Group by: split-apply-combine

Another operation worth learning is how to aggregate the data based on some criteria. The method we are going to use is called `groupby()`, and is actually the result of combining the following steps:

- **splitting** the data into groups based on some criteria
- **applying** a function to each group independently
- **combining** the results into a data structure

To demonstrate this, we will use the following dataset containing information on passengers of the Titanic.

In [15]:
titanic = pd.read_csv('data/titanic.csv')

This dataset contains infomation on the passengers, like the sex, the name, the age, and some info about the ticket prices and so on

In [16]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Suppose we want to make some statistics on the age of the passengers.
One way we could do it is by selecting the 'Age' column:


In [17]:
ages = titanic['Age']

In [18]:
ages

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

Or we could select both age and sex of the passengers, and obtain a new DataFrame:

In [19]:
age_sex = titanic[['Age', 'Sex']]

In [20]:
age_sex.head(10)

Age     Sex
0  22.0    male
1  38.0  female
2  26.0  female
3  35.0  female
4  35.0    male
5   NaN    male
6  54.0    male
7   2.0    male
8  27.0  female
9  14.0  female

Now, suppose we want to calculate the average age, we could simpy select the 'Age' column and calculate the mean:

In [21]:
ages.mean()

29.69911764705882

But if we wanted to calculate the average age depending on the sex, we cannot calculate it as:

In [31]:
titanic[['Age', 'Sex']].mean(numeric_only=True)

Age    29.699118
dtype: float64

As you can see, it doesn't give us what we were expecting. This happens because 'Sex' is a cathegorical variable, that can have the values "male" or "female", and it isn't understood by the mean function. With two numerical variables, like the age and the ticket fare, we could do:

In [23]:
titanic[['Age', 'Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

And this would give us the mean of each column.

What if we wanted to know the average age for each sex?

To do this we can use the `groupby()` method, to make a group per category:

In [24]:
titanic.groupby('Sex').mean()

/var/folders/r8/fr9gwf115c5cy6kt67ntkq7h0000gp/T/ipykernel_8098/662762107.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  titanic.groupby('Sex').mean()


PassengerId  Survived    Pclass        Age     SibSp     Parch  \
Sex                                                                      
female   431.028662  0.742038  2.159236  27.915709  0.694268  0.649682   
male     454.147314  0.188908  2.389948  30.726645  0.429809  0.235702   

             Fare  
Sex                
female  44.479818  
male    25.523893

As expected, this returns the mean of each column, grouped according to 'Sex'. Let's select the 'Age':

In [25]:
titanic.groupby('Sex')['Age'].mean()

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

It is possible to group data by more categories at the same time. So, if we wanted to know the average cost of a ticket for both genders and for different cabin classes:

In [26]:
titanic.groupby(['Sex', 'Pclass'])['Fare'].mean()

Sex     Pclass
female  1         106.125798
        2          21.970121
        3          16.118810
male    1          67.226127
        2          19.741782
        3          12.661633
Name: Fare, dtype: float64

### Count the number of records per Category

What is the number of passengers in each cabin class?

In [27]:
titanic['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

The `value_counts()` method counts the number of records for each category in a column. The function is a shortcut, as it is actually a groupby operation in combination with counting of the number of records within each group:

In [28]:
titanic.groupby('Pclass')['Pclass'].count()

Pclass
1    216
2    184
3    491
Name: Pclass, dtype: int64

Both size and count can be used in combination with groupby. Whereas size includes `NaN` values and just provides the number of rows (size of the table), count excludes the missing values. In the `value_counts` method, use the `dropna` argument to include or exclude the `NaN`
 values. more info: https://pandas.pydata.org/docs/user_guide/basics.html#basics-discretization

REMEMBER
- Aggregation statistics can be calculated on entire columns or rows
- `groupby` provides the power of the split-apply-combine pattern
- `value_counts`
 is a convenient shortcut to count the number of entries in each category of a variable

These are some of the most common operations we can perform on a Pandas DataFrame. 
For more details, you can always refer to the official documentation, that also includes very detailed tutorials:https://pandas.pydata.org/docs/getting_started/intro_tutorials/index.html